README.md
# PoTi
PoTi: Pomodor Timer 

OS: Windows, Linux


CHANGELOG.md
# v0.4.2
# Fixes

1. Fix user_data folder creation

# v0.4.1

## Changes

1. Fix: bell now ring
2. Change library to play sound

# v0.4

## Changes

1. Move to customtkinter library
2. Improve project structure

# v0.3

## Changes

1. Add new window with settings
2. Now statistics were saved on every closed sessions
3. Reset button only resets stages, it doens't remove closed sessions mark
4. On start up, program load work sessions and display certain amount of tick

## Fixes

1. Statistics save on close app
2. If completed sessions equal 0 - this not will be save
3. Program content now centered

# v0.3-pre.2

## Changes

1. Now statistics were saved on every closed sessions
2. Reset button only resets stages,it doens't remove closed sessions mark
3. On start up, program load work sessions and display certain amount of tick

# v0.3-pre.1

## Fixes

1. Statistics save on close app
2. If completed sessions equal 0 - this not will be save
3. Program content now centered

# v0.2.3

## Fixes

1. Tick's row doesn't update on screen

# v0.2.2

## Changes

1. Add all configuration options to **_settings.json_**
2. Clean code, improve file structure

## Fixes

1. I guess PoTi will update 'stats.json' with new date

# v0.2.1

## Changes

1. In 'settings.json' you can set duration of work, break and long break stages

## Fixes

1. Now you can run program from any path
2. File 'stats.json' update correctly

# v0.2

## Changes

1. Add ring bell when stage are end
2. Now statistics save in stats.json


ui.py
from tkinter import *
from tkinter import messagebox
from timer_process import TimerProcces
from playsound import playsound


class PoTiInterface:
    def __init__(self, config: dict, timer_process: TimerProcces) -> None:
        self.config = config

        self.timer = None
        self.process = timer_process

        self.window = Tk()
        self.window.geometry("600x450")
        self.window.minsize(600, 450)
        self.window.title("Pomodoro")
        self.window.config(padx=100, pady=30, bg=self.config["background_color"])

        self.frame = Frame(bg=self.config["background_color"])
        self.canvas = Canvas(
            self.frame,
            width=200,
            height=223,
            highlightthickness=0,
            bg=self.config["background_color"],
        )
        self.tomato_image = PhotoImage(file=self.config["image"])
        self.canvas.create_image(100, 111, image=self.tomato_image)
        self.timer_text = self.canvas.create_text(
            103,
            130,
            text="00:00",
            fill="white",
            font=(self.config["font_name"], 29, "bold"),
        )
        self.canvas.grid(column=1, row=1)

        self.name_label = Label(
            self.frame,
            text="Timer",
            fg=self.config["foreground_color"],
            bg=self.config["background_color"],
            font=(self.config["font_name"], 38),
        )
        self.name_label.grid(column=1, row=0)
        self.name_label.config(pady=10)

        self.start_button = Button(
            self.frame,
            text="Start",
            height=1,
            width=8,
            bg=self.config["background_color"],
            font=(self.config["font_name"], 13, "bold"),
            command=self.start_timer,
        )
        self.start_button.grid(column=0, row=2)
        self.start_button.config(pady=10)

        self.reset_button = Button(
            self.frame,
            text="Reset",
            height=1,
            width=8,
            bg=self.config["background_color"],
            font=(self.config["font_name"], 13, "bold"),
            command=self.reset_timer,
        )
        self.reset_button.grid(column=2, row=2)
        self.reset_button.config(pady=10)

        self.settings_button = Button(
            self.frame,
            text="⚙",
            height=1,
            width=1,
            bg=self.config["background_color"],
            font=(self.config["font_name"], 13, "bold"),
            command=self.settings,
        )
        self.settings_button.place(relx=0.85, rely=0.01)

        self.tick_label = Label(
            self.frame,
            bg=self.config["background_color"],
            fg=self.config["foreground_color"],
            font=(self.config["font_name"], 16),
        )
        self.tick_label.grid(column=1, row=3)
        self.tick_label.config(text=self.process.mark)
        self.frame.place(anchor=CENTER, relx=0.5, rely=0.5)

        self.window.mainloop()

    def reset_timer(self):
        self.process.reset_timer()
        self.window.after_cancel(self.timer)
        self.canvas.itemconfig(self.timer_text, text="00:00")
        self.name_label.config(text="Timer")
        self.tick_label.config(text="")

    def start_timer(self):
        work_sec = self.config["work_min"] * 60
        short_break = self.config["break_min"] * 60
        long_break = self.config["long_break_min"] * 60

        self.process.reps += 1

        if self.process.reps % 8 == 0:
            self.name_label.config(
                text="Long Break", fg=self.config["long_break_font_color"]
            )
            self.count_down(long_break)
        elif self.process.reps % 2 != 0:
            self.name_label.config(text="Work", fg=self.config["foreground_color"])
            self.count_down(work_sec)
        else:
            self.name_label.config(text="Break", fg=self.config["break_font_color"])
            self.count_down(short_break)

    def count_down(self, count):
        count_min = count // 60
        count_sec = count % 60
        if count_sec < 10:
            count_sec = f"0{count_sec}"

        self.canvas.itemconfig(self.timer_text, text=f"{count_min}:{count_sec}")
        if count > 0:
            self.timer = self.window.after(1000, self.count_down, count - 1)
        else:
            playsound(self.config["sound"])
            self.start_timer()
            if self.process.reps % 2 == 0:
                self.process.mark = f"{self.process.mark}" + "✓"
                self.tick_label.config(text=self.process.mark)
                self.process.update_stats()

    def settings(self):
        def change_config():
            self.config["work_min"] = int(work_min_entry.get())
            self.config["break_min"] = int(break_min_entry.get())
            self.config["long_break_min"] = int(long_break_min_entry.get())
            self.config["background_color"] = bg_color_entry.get()
            self.config["foreground_color"] = fg_color_entry.get()
            self.config["break_font_color"] = break_font_color_entry.get()
            self.config["long_break_font_color"] = long_break_font_color_entry.get()
            self.config["font_name"] = font_name_entry.get()
            self.process.save_config()
            settings_window.destroy()
            messagebox.showinfo(
                title="Done",
                message="Settings are applied",
            )

        def change_default():
            self.config["work_min"] = 25
            self.config["break_min"] = 5
            self.config["long_break_min"] = 15
            self.config["background_color"] = "#d3eca7"
            self.config["foreground_color"] = "#a1b57d"
            self.config["break_font_color"] = "#19282f"
            self.config["long_break_font_color"] = "#eb3303"
            self.config["font_name"] = "Courier"
            self.process.save_config()
            settings_window.destroy()
            messagebox.showinfo(
                title="Done",
                message="Default settings are restored",
            )

        settings_window = Toplevel(self.window)
        settings_window.title("PoTi Settings")
        settings_window.geometry("550x430")
        settings_window.resizable(False, False)
        settings_window.config(padx=20, pady=20, bg=self.config["background_color"])

        work_min_label = Label(
            settings_window,
            text="Work duration(min): ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        work_min_label.grid(column=0, row=2, sticky="e")
        work_min_label.config(pady=10)
        work_min_entry = Entry(settings_window, width=15)
        work_min_entry.insert(0, self.config["work_min"])
        work_min_entry.grid(column=1, row=2)

        break_min_label = Label(
            settings_window,
            text="Break duration(min): ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        break_min_label.grid(column=0, row=3, sticky="e")
        break_min_label.config(pady=10)
        break_min_entry = Entry(settings_window, width=15)
        break_min_entry.insert(0, self.config["break_min"])
        break_min_entry.grid(column=1, row=3)

        long_break_min_label = Label(
            settings_window,
            text="Long break duration(min): ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        long_break_min_label.grid(column=0, row=4, sticky="e")
        long_break_min_label.config(pady=10)
        long_break_min_entry = Entry(settings_window, width=15)
        long_break_min_entry.insert(0, self.config["long_break_min"])
        long_break_min_entry.grid(column=1, row=4)

        bg_color_label = Label(
            settings_window,
            text="Background color: ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        bg_color_label.grid(column=0, row=0, sticky="e")
        bg_color_label.config(pady=10)
        bg_color_entry = Entry(settings_window, width=15)
        bg_color_entry.insert(0, self.config["background_color"])
        bg_color_entry.grid(column=1, row=0)

        fg_color_label = Label(
            settings_window,
            text="Foreground color: ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        fg_color_label.grid(column=0, row=1, sticky="e")
        fg_color_label.config(pady=10)
        fg_color_entry = Entry(settings_window, width=15)
        fg_color_entry.insert(0, self.config["foreground_color"])
        fg_color_entry.grid(column=1, row=1)

        break_font_color_label = Label(
            settings_window,
            text="Break font color: ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        break_font_color_label.grid(column=0, row=5, sticky="e")
        break_font_color_label.config(pady=10)
        break_font_color_entry = Entry(settings_window, width=15)
        break_font_color_entry.insert(0, self.config["break_font_color"])
        break_font_color_entry.grid(column=1, row=5)

        long_break_font_color_label = Label(
            settings_window,
            text="Long break font color: ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        long_break_font_color_label.grid(column=0, row=6, sticky="e")
        long_break_font_color_label.config(pady=10)
        long_break_font_color_entry = Entry(settings_window, width=15)
        long_break_font_color_entry.insert(0, self.config["long_break_font_color"])
        long_break_font_color_entry.grid(column=1, row=6)

        font_name_label = Label(
            settings_window,
            text="Font family name: ",
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
        )
        font_name_label.grid(column=0, row=7, sticky="e")
        font_name_label.config(pady=10)
        font_name_entry = Entry(settings_window, width=15)
        font_name_entry.insert(0, self.config["font_name"])
        font_name_entry.grid(column=1, row=7)

        apply_btn = Button(
            settings_window,
            text="Apply",
            height=1,
            width=8,
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
            command=change_config,
        )
        apply_btn.grid(column=0, row=8, pady=10)
        reset_default_btn = Button(
            settings_window,
            text="Default",
            height=1,
            width=8,
            bg=self.config["background_color"],
            font=(self.config["font_name"], 14, "bold"),
            command=change_default,
        )
        reset_default_btn.grid(column=1, row=8, pady=10)


poti/stats.py
from datetime import datetime
import json
from json import JSONDecodeError
import poti.timer_process as ptp
from poti.config import config


def get_data():
        try:
            with open(config.config["stats"], "r") as stats:
                data = json.load(stats)
                today = datetime.now().strftime("%Y-%m-%d")
                sessions = data[today]["work_sessions"]
                for _ in range(1, sessions + 1):
                    ptp.mark = ptp.mark + "✓"

                return ptp.mark
        except (JSONDecodeError, FileNotFoundError, KeyError):
            pass


def update_stats():
        if len(ptp.mark) != 0:
            today = datetime.now()
            today = today.strftime("%Y-%m-%d")
            work_sessions = len(ptp.mark)
            overal_hours = work_sessions * config.config["work_min"] // 60
            overal_mins = work_sessions * config.config["work_min"] % 60
            if overal_mins < 10:
                overal_mins = f"0{overal_mins}"
            overal_time = f"{overal_hours}:{overal_mins}"
            new_data = {
                today: {
                    "overall_time": overal_time,
                    "work_sessions": work_sessions,
                }
            }
            try:
                with open(config.config["stats"], "r") as stats:
                    data = json.load(stats)

            except (JSONDecodeError, FileNotFoundError):
                with open(config.config["stats"], "w") as stats:
                    json.dump(new_data, stats, indent=4)
            else:
                if today in data:
                    work_sessions = len(ptp.mark)
                    overal_hours = work_sessions * config.config["work_min"] // 60
                    overal_mins = work_sessions * config.config["work_min"] % 60
                    if overal_mins < 10:
                        overal_mins = f"0{overal_mins}"
                    overal_time = f"{overal_hours}:{overal_mins}"
                    new_data = {
                        today: {
                            "overall_time": overal_time,
                            "work_sessions": work_sessions,
                        }
                    }
                data.update(new_data)
                with open(config.config["stats"], "w") as stats:
                    json.dump(data, stats, indent=4)


poti/poti_ui.py
import tkinter
import customtkinter

import poti.timer_process as ptp
from poti.stats import get_data
from poti.poti_settings_ui import PoTiSettings



# customtkinter.set_appearance_mode("System")  # Modes: "System" (standard), "Dark", "Light"
# customtkinter.set_default_color_theme("blue")  # Themes: "blue" (standard), "green", "dark-blue"


class PoTiUi(customtkinter.CTk):
    def __init__(self):
        super().__init__()

        self.timer = None

        self.title("PoTi")
        self.geometry(f"{960}x{540}")
        self.minsize(960, 540)
        self.configure(padx=100, pady=30)

        self.main_frame = customtkinter.CTkFrame(self, width=960, height=540, fg_color=self._fg_color)
        self.main_frame.place(relx=0.5, rely=0.5, anchor=tkinter.CENTER)

        self.stage_status = customtkinter.CTkLabel(
            self.main_frame,
            text="Timer",
            font=customtkinter.CTkFont(size=38)
        )
        self.stage_status.grid(column=1, row=0)

        self.timer_countdown = customtkinter.CTkLabel(
            self.main_frame,
            text="00:00",
            font=customtkinter.CTkFont(size=72)
        )
        self.timer_countdown.grid(column=1, row=1)
        self.timer_countdown.configure(pady=70)

        self.start_button = customtkinter.CTkButton(
            self.main_frame,
            text="Start",
            width=120,
            height=50,
            command=ptp.start_timer
        )
        self.start_button.grid(row=2, column=0)

        self.reset_button = customtkinter.CTkButton(
            self.main_frame,
            text="Reset",
            width=120,
            height=50,
            command=ptp.reset_timer
        )
        self.reset_button.grid(row=2, column=2)

        self.tick_label = customtkinter.CTkLabel(
            self.main_frame,
            text=get_data(),
            font=customtkinter.CTkFont(size=24),
        )
        self.tick_label.grid(row=3, column=1)

        self.settings_button = customtkinter.CTkButton(
            self.main_frame,
            text="⚙",
            height=24,
            width=24,
            font=customtkinter.CTkFont(size=20),
            command=self.settings
        )
        self.settings_button.place(relx=0.85, rely=0.01)

        self.toplevel_window = None

    def settings(self):
        self.toplevel_window = PoTiSettings()



app = PoTiUi()
app.mainloop()


poti/timer_process.py
import pygame

import poti.poti_ui as pu
from poti.config import config
from poti.stats import update_stats


reps = 0
mark = ""

pygame.mixer.init()

def start_timer():
    work_sec = config.config['work_min'] * 60
    short_break = config.config['short_break_min'] * 60
    long_break = config.config['long_break_min'] * 60

    global reps
    reps += 1

    if reps % 8 == 0:
        pu.app.stage_status.configure(
            text="Long Break"
        )
        count_down(long_break)
    elif reps % 2 != 0:
        pu.app.stage_status.configure(text="Work")
        count_down(work_sec)
    else:
        pu.app.stage_status.configure(text="Break")
        count_down(short_break)


def count_down(seconds):
    '''Run countdown'''
    count_min = seconds // 60
    count_sec = seconds % 60
    if count_sec < 10:
        count_sec = f"0{count_sec}"

    pu.app.timer_countdown.configure(text=f"{count_min}:{count_sec}")
    if seconds > 0:
        pu.app.timer = pu.app.after(1000, count_down, seconds - 1)
    else:
        pygame.mixer.music.load(config.config['sound'])
        pygame.mixer.music.play(loops=0)
        start_timer()

        global reps
        if reps % 2 == 0:
            global mark
            mark += "✓"
            pu.app.tick_label.configure(text=mark)
            update_stats()


def reset_timer():
    global reps
    reps = 0

    pu.app.after_cancel(pu.app.timer)
    pu.app.timer_countdown.configure(text="00:00")
    pu.app.stage_status.configure(text="Timer")
    pu.app.tick_label.configure(text="")



poti/poti_settings_ui.py
import customtkinter
from CTkMessagebox import CTkMessagebox

from poti.config import config


class PoTiSettings(customtkinter.CTkToplevel):
    def __init__(self):
        super().__init__()

        self.title("PoTi Settings")
        # self.geometry("550x430")
        self.resizable(False, False)
        self.config(padx=20, pady=20)

        self.work_min_label = customtkinter.CTkLabel(
            self,
            text="Work duration(min): ",
        )
        self.work_min_label.grid(column=0, row=2, sticky="e")
        self.work_min_label.configure(pady=10)

        self.work_min_entry = customtkinter.CTkEntry(self, width=30)
        self.work_min_entry.insert(0, string=config.config["work_min"])
        self.work_min_entry.grid(column=1, row=2)

        self.break_min_label = customtkinter.CTkLabel(
            self,
            text="Break duration(min): ",
        )
        self.break_min_label.grid(column=0, row=3, sticky="e")
        self.break_min_label.configure(pady=10)

        self.break_min_entry = customtkinter.CTkEntry(self, width=30)
        self.break_min_entry.insert(0, string=config.config["short_break_min"])
        self.break_min_entry.grid(column=1, row=3)

        self.long_break_min_label = customtkinter.CTkLabel(
            self,
            text="Long break duration(min): ",
        )
        self.long_break_min_label.grid(column=0, row=4, sticky="e")
        self.long_break_min_label.configure(pady=10)

        self.long_break_min_entry = customtkinter.CTkEntry(self, width=30)
        self.long_break_min_entry.insert(0, string=config.config["long_break_min"])
        self.long_break_min_entry.grid(column=1, row=4)

        self.apply_btn = customtkinter.CTkButton(
            self,
            text="Apply",
            height=1,
            width=8,
            command=self.change_config,
        )
        self.apply_btn.grid(column=0, row=8, pady=10)

        self.reset_default_btn = customtkinter.CTkButton(
            self,
            text="Default",
            height=1,
            width=8,
            command=self.change_default,
        )
        self.reset_default_btn.grid(column=1, row=8, pady=10)

    def change_config(self):
               config.config["work_min"] = int(self.work_min_entry.get())
               config.config["short_break_min"] = int(self.break_min_entry.get())
               config.config["long_break_min"] = int(self.long_break_min_entry.get())
               config.save_config(config)

               self.destroy()
               CTkMessagebox(
                   title="Done",
                   message="Settings are applied"
               )

    def change_default(self):
        config.config["work_min"] = 25
        config.config["short_break_min"] = 5
        config.config["long_break_min"] = 15
        config.save_config(config)

        self.destroy()
        CTkMessagebox(
            title="Done",
            message="Default settings are restored"
        )


poti/config.py
import json
import os
import sys

if getattr(sys, "frozen", False):
    application_path = os.path.dirname(sys.executable)
else:
    application_path = os.getcwd()

USER_DATA_FOLDER = os.path.join(application_path, "user_data")
RESOURCES_FOLDER = os.path.join(application_path, "resources")

if os.path.exists(USER_DATA_FOLDER):
    pass
else:
    os.makedirs(USER_DATA_FOLDER)

SOUND_PATH = os.path.join(RESOURCES_FOLDER, "bell.wav")
STATS_PATH = os.path.join(USER_DATA_FOLDER, "stats.json")
CONFIG_PATH = os.path.join(USER_DATA_FOLDER, "settings.json")

def get_config():
        config = {}
        try:
            with open(CONFIG_PATH, "r") as settings:
                config = json.load(settings)
        except FileNotFoundError:
            with open(CONFIG_PATH, "w") as settings:
                config = {
                    "work_min": 25,
                    "short_break_min": 5,
                    "long_break_min": 15,
                }
                json.dump(config, settings, indent=4)
        finally:
            config["sound"] = SOUND_PATH
            config["stats"] = STATS_PATH
        return config

class PoTiConfig():
    def __init__(self, startup_config: dict):
        self.config = startup_config

    def save_config(self, config):
       with open(CONFIG_PATH, "w") as config:
           new_config = {
               "work_min": self.config["work_min"],
               "short_break_min": self.config["short_break_min"],
               "long_break_min": self.config["long_break_min"],
           }
           json.dump(new_config, config, indent=4)

startup_config = get_config()
config = PoTiConfig(startup_config)
